staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go

```diff
@@ -0,0 +1,192 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldmanager
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+)
+
+var (
+	scaleGroupVersion = schema.GroupVersion{Group: "autoscaling", Version: "v1"}
+	scalePath         = fieldpath.MakePathOrDie("spec", "replicas")
+)
+
+// SubresourceScaler manages the conversion of managed fields between a main
+// resource and subresources
+type SubresourceScaler struct {
+	parentEntries      []metav1.ManagedFieldsEntry
+	parentGroupVersion schema.GroupVersion
+	parentPath         fieldpath.Path
+}
+
+// ScalePathMappings maps a group/version to the replicas path in a scale subresource
+type ScalePathMappings map[schema.GroupVersion]fieldpath.Path
+
+// NewSubresourceScaler creates a new SubresourceScaler
+func NewSubresourceScaler(parentEntries []metav1.ManagedFieldsEntry, parentGroupVersion schema.GroupVersion, parentPath fieldpath.Path) (*SubresourceScaler, error) {
+	return &SubresourceScaler{
+		parentEntries:      parentEntries,
+		parentGroupVersion: parentGroupVersion,
+		parentPath:         parentPath,
+	}, nil
+}
+
+// ToSubresource filter the managed fields of the main resource and convert
+// them so that they can be handled by a subresource.
+// For the managed fields that have a path that could be handled by a
+// subresource this method performs two changes:
+// 1. The APIVersion is changed to the APIVersion of the subresource
+// 2. The Fields are transformed to the path of that subresource
+func (h *SubresourceScaler) ToSubresource() ([]metav1.ManagedFieldsEntry, error) {
+	managed, err := internal.DecodeObjectManagedFields(h.parentEntries)
+	if err != nil {
+		return nil, err
+	}
+
+	f := fieldpath.ManagedFields{}
+	t := map[string]*metav1.Time{}
+	for manager, versionedSet := range managed.Fields() {
+
+		if versionedSet.Set().Has(h.parentPath) {
+			newVersionedSet := fieldpath.NewVersionedSet(
+				fieldpath.NewSet(scalePath),
+				fieldpath.APIVersion(scaleGroupVersion.String()),
+				versionedSet.Applied(),
+			)
+
+			f[manager] = newVersionedSet
+			if operationTime, ok := managed.Times()[manager]; ok {
+				t[manager] = operationTime
+			}
+		}
+
+	}
+
+	return managedFieldsEntries(internal.NewManaged(f, t))
+}
+
+// ToParent merges the `subresourceEntries` with the parent entries and
+// transform them accordingly
+func (h *SubresourceScaler) ToParent(scaleEntries []metav1.ManagedFieldsEntry) ([]metav1.ManagedFieldsEntry, error) {
```

Comment:
Still not sure I agree with the algorithm here, here's what I can see:
1. There is a new value for replicas, whoever owned it before won't anymore, new manager is inserted
2. The value of replicas is the same as before, everyone should still be able to own their value
a. If the value was applied, we need a new manager for that apply, as in 1.
b. if the value was updated, I'm curious if we should add it anyway.

However, I don't know if that's the most convincing way to implement it.

Have you tried something like this:
- For each entry in parentsManagedFields that has replicas field,
  - Search corresponding entry in scaleManagedFields
    - If not found, remove the field from parent entry (it lost the field)
    - if found, remove that entry from scaleManagedFields (it kept the field)
- For each entry in scaleManagedFields left (assert that there should be one at most, newly added replicas owner)
  - Add new entry in parentsManagedFields of default version
- Return parentsManagedFields